clients/client-go/tcobject/upload.go

```diff
@@ -0,0 +1,160 @@
+
+	uploadID := slugid.Nice()
+	proposedUploadMethods := ProposedUploadMethods{}
+
+	if contentLength < DataInlineMaxSize {
+		content, err := ioutil.ReadAll(readSeeker)
+		if err != nil {
+			return err
+		}
+		b64 := base64.StdEncoding.EncodeToString(content)
+		proposedUploadMethods.DataInline = DataInlineUploadRequest{
+			ContentType: contentType,
+			ObjectData:  b64,
+		}
+	}
+
+	proposedUploadMethods.PutURL = PutURLUploadRequest{
+		// TODO: find out why this is float64 when it should be int64
+		ContentLength: float64(contentLength),
+		ContentType:   contentType,
+	}
+
+	uploadResp, err := object.CreateUpload(
+		name,
+		&CreateUploadRequest{
+			Expires:               tcclient.Time(expires),
+			ProjectID:             projectID,
+			UploadID:              uploadID,
+			ProposedUploadMethods: proposedUploadMethods,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Ensure FinishUpload is called even if upload fails by placing in a defer
+	// block.
+	defer func() {
+		err2 := object.FinishUpload(
+			name,
+			&FinishUploadRequest{
+				ProjectID: projectID,
+				UploadID:  uploadID,
+			},
+		)
+		// only update returned error if it isn't already set
+		if err == nil {
+			err = err2
+		}
+	}()
+
+	switch {
+	case uploadResp.UploadMethod.DataInline:
+		// data is already uploaded -- nothing to do
+		return nil
+	case uploadResp.UploadMethod.PutURL.URL != "":
+		return putURLUpload(object.HTTPBackoffClient, uploadResp.UploadMethod, readSeeker)
+	}
+	return errors.New("Could not negotiate an upload method")
+}
+
+func putURLUpload(httpBackoffClient *httpbackoff.Client, uploadMethod SelectedUploadMethodOrNone, readSeeker io.ReadSeeker) error {
+	// perform http PUT to upload to S3...
+	httpClient := &http.Client{}
+	httpCall := func() (putResp *http.Response, tempError error, permError error) {
+		// Explicitly seek to start here, rather than only after a temp error,
+		// since not all temporary errors are caught by this code (e.g. status
+		// codes 500-599 are handled by httpbackoff library implicitly).
+		_, permError = readSeeker.Seek(0, io.SeekStart)
+		if permError != nil {
+			// not being able to seek to start is a problem that is unlikely to
+			// be solved with retries with exponential backoff, so give up
+			// straight away
+			return
+		}
+		var httpRequest *http.Request
+		httpRequest, permError = http.NewRequest("PUT", uploadMethod.PutURL.URL, readSeeker)
+		if permError != nil {
+			return
+		}
+		for headerName, headerValue := range uploadMethod.PutURL.Headers {
+			httpRequest.Header.Set(headerName, headerValue)
+		}
+		putResp, tempError = httpClient.Do(httpRequest)
+		if tempError != nil {
+			return
+		}
+		// bug 1394557: s3 incorrectly returns HTTP 400 for connection
+		// inactivity, so make this a temp error rather than a permanent error
+		// if it occurs, in case this is a put request to s3
```

Comment:
Can you add an issue to do the same in the other clients?

Reply:
Created #4623